File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/MapChunkBulkPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import java.util.List;

import gg.mineral.api.world.chunk.Chunk;
import gg.mineral.api.network.packet.Packet;
import gg.mineral.server.world.chunk.ChunkImpl;
import io.netty.buffer.ByteBuf;
import lombok.val;

public final record MapChunkBulkPacket(boolean skyLight, List<Chunk> chunks) implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        int amount = this.chunks.size();
        val bytes = new byte[196864 * amount];
        int bytesPosition = 0;

        val output = new Pair[amount];
        for (int i = 0; i < amount; i++) {
            if (this.chunks.get(i) instanceof ChunkImpl chunk) {
                val out = chunk.toPacket(skyLight, false);
                System.arraycopy(out.compressedData(), 0, bytes, bytesPosition, out.compressedData().length);
                bytesPosition += out.compressedData().length;
                output[i] = new Pair(chunk, out);
            }
        }

        val compressed = ChunkImpl.compress(bytes, bytesPosition);

        os.writeShort(amount);
        os.writeInt(compressed.length);
        os.writeBoolean(skyLight);
        os.writeBytes(compressed);

        for (val pair : output) {
            val col = pair.chunk();
            val out = pair.packet();
            os.writeInt(col.getX());
            os.writeInt(col.getZ());
            os.writeShort(out.primaryBitMap());
            os.writeShort(out.addBitMap());
        }
    }

    private record Pair(Chunk chunk, ChunkDataPacket packet) {
    }

    @Override
    public byte getId() {
        return 0x26;
    }

}

File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/MapChunkBulkPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import gg.mineral.api.world.chunk.Chunk
import gg.mineral.api.world.chunk.Chunk.x
import gg.mineral.api.world.chunk.Chunk.z
import gg.mineral.server.world.chunk.ChunkImpl
import gg.mineral.server.world.chunk.ChunkImpl.Companion.compress
import gg.mineral.server.world.chunk.ChunkImpl.toPacket
import io.netty.buffer.ByteBuf

@JvmRecord
data class MapChunkBulkPacket(val skyLight: Boolean, val chunks: List<Chunk>) :
    Packet.OUTGOING {
    override fun serialize(os: ByteBuf) {
        val amount = chunks.size
        val bytes = ByteArray(196864 * amount)
        var bytesPosition = 0

        val output = arrayOfNulls<Pair>(amount)
        for (i in 0..<amount) {
            if (chunks[i] is ChunkImpl) {
                val out: ChunkDataPacket = chunk.toPacket(skyLight, false)
                System.arraycopy(out.compressedData, 0, bytes, bytesPosition, out.compressedData.size)
                bytesPosition += out.compressedData.size
                output[i] = Pair(chunk, out)
            }
        }

        val compressed = compress(bytes, bytesPosition)

        os.writeShort(amount)
        os.writeInt(compressed.size)
        os.writeBoolean(skyLight)
        os.writeBytes(compressed)

        for ((col, out) in output) {
            os.writeInt(col.x.toInt())
            os.writeInt(col.z.toInt())
            os.writeShort(out.primaryBitMap)
            os.writeShort(out.addBitMap)
        }
    }

    override fun getId(): Byte {
        return 0x26
    }

    @JvmRecord
    private data class Pair(val chunk: Chunk, val packet: ChunkDataPacket)
}
